### PR TITLE
Task/cache hooks

### PR DESCRIPTION
**Pull request by sav007**
This PR is a proposal of adding hooks to all API calls that will be called in order to cache response got back. The idea is to help user not to wrap each API call that he/she wants to cache on client side but to notify through hooks when and what needs to be cached. As well we are giving some contract what should be cached. To inject specific hook user should add particular implementation of it during building BuyClient. 

There are several interfaces that specifies contract for the hooks:
- AddressCacheHook for caching responses related to customer address APIs
- CheckoutCacheHook for caching responses related to customer checkout APIs 
- CustomerCacheHook for caching responses related to customer APIs 
- OrderCacheHook for caching responses related to customer order APIs 
- ProductCacheHook for caching responses related to product listing APIs 
- StoreCacheHook for caching responses related to shop APIs 

All cache hook calls are operated at the same thread as BuyClient API call.

@krisorr @bgulanowski

### CODE REVIEW

**Comment by bgulanowski** [comment]
@sav007 if you have time, could you add something to the description for this PR? For example:
- pointer to issue describing what this resolves
  - if there is no issue, please write one up
- pointer to what CacheHooks are and where they come from (I see most of this is code brought in from elsewhere; just want to know where from)
- discussion of why this is raw source, instead of library dependency; if no lib exists, should we make one?
- how this functionality integrates into the SDK

Thanks.

**Comment by sav007** [comment]
@bgulanowski confused about this:

pointer to what CacheHooks are and **where they come from (I see most of this is code brought in from elsewhere; just want to know where from)**

what do you mean exactly? it's not a library it's implemented from scratch.

**Comment by bgulanowski** [comment]
@sav007 my mistake! I didn't realize this was the SDK, I thought it was an internal project. So when I saw the MIT license, and how much code there was, I assumed you brought in a library. But it's our library!

Pretty big change, so it will take a while to review.

**Comment by sav007** [comment]
It's not so big as seems, a lot of code for unit tests, but everything else is straight forward.

File: MobileBuy/buy/src/main/java/com/shopify/buy/dataprovider/cache/CheckoutCacheHook.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+import com.shopify.buy.model.Checkout;
+import com.shopify.buy.model.ShippingRate;
+
+import java.util.List;
+
+/**
+ * Cache hook that will be triggered by {@link com.shopify.buy.dataprovider.CheckoutService}. By default all caching operates
+ * on background thread.
+ */
+public interface CheckoutCacheHook {
```
Comment (bgulanowski) [inline MobileBuy/buy/src/main/java/com/shopify/buy/dataprovider/cache/CheckoutCacheHook.java]: Checkouts should not be cached. Only carts.

**Comment by bgulanowski** [comment]
I gave this a look, but I still can't infer all the implications and consequences. An overview would be helpful, but it's not urgent. This will have to wait for Kris to review thoroughly.

File: MobileBuy/buy/src/main/java/com/shopify/buy/dataprovider/cache/CheckoutCacheHook.java (hunk 1 above)
Comment (sav007) [inline MobileBuy/buy/src/main/java/com/shopify/buy/dataprovider/cache/CheckoutCacheHook.java]: Yeah I thought about this, the issue that for Great app we are using checkout as a cart that seems not right, well we will go over and discuss what we need to remove from caching.

**Pull request closed without merging**